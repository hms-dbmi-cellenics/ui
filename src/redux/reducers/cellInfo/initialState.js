const initialState = {
  component: null,
  cellId: null,
  geneName: null,
  trackCluster: null,
  focus: {
    store: 'cellSets',
    key: 'louvain',
  },
  groupedTrack: 'louvain',
  selectedTracks: ['louvain'],
};

export default initialState;
