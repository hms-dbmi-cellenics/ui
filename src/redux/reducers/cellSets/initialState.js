const initialState = {
  properties: {},
  hierarchy: [],
  selected: {},
  initialLoadPending: true,
  loading: false,
  error: false,
  updatingClustering: false,
  hidden: new Set(),
};

export default initialState;
