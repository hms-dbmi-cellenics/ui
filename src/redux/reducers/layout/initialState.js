const initialLayoutSpatial = {
  windows: {
    direction: 'row',
    first: {
      first: {
        first: 'UMAP', second: 'Spatial', direction: 'row', splitPercentage: 62,
      },
      second: {
        first: 'Heatmap',
        second: 'Cell sets and Metadata',
        direction: 'row',
        splitPercentage: 62,
      },
      direction: 'column',
      splitPercentage: 45,
    },
    second: 'Genes',
    splitPercentage: 70,
  },
  panel: 'Gene list',
};

const initialLayoutSingleCell = {
  windows: {
    direction: 'row',
    first: {
      first: {
        first: 'UMAP', second: 'Cell sets and Metadata', direction: 'row', splitPercentage: 62,
      },
      second: 'Heatmap',
      direction: 'column',
      splitPercentage: 45,
    },
    second: 'Genes',
    splitPercentage: 70,
  },
  panel: 'Gene list',
};

export { initialLayoutSpatial, initialLayoutSingleCell };
