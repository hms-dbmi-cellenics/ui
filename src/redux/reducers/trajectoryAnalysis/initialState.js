const initialState = {
  loading: false,
  error: false,
  plotData: {},
};

export default initialState;
