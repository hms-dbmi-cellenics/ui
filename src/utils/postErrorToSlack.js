import Auth from '@aws-amplify/auth';
import StackTrace from 'stacktrace-js';
import { getLoggerBotToken } from 'utils/slack';

const extractExperimentId = (url) => {
  const match = url.match(/experiments\/([^/]+)/i);
  return match ? match[1] : null;
};

// Truncates data arrays so that it doesn't produce large logs
const NUM_DATA_TO_SHOW = 20;
const truncateCollection = (arr) => {
  if (arr.length < NUM_DATA_TO_SHOW) return arr;

  const truncatedArr = arr.slice(0, NUM_DATA_TO_SHOW);
  truncatedArr.push('...');
  return truncatedArr;
};

const trimOutput = (key, item) => {
  if (Array.isArray(item)) return truncateCollection(item);
  if (item instanceof Set) return new Set(truncateCollection(Array.from(item)));

  if (item !== null && typeof item === 'object') {
    const newItem = Object.keys(item).reduce((newObject, childKey) => {
      // eslint-disable-next-line no-param-reassign
      newObject[childKey] = trimOutput(childKey, item[childKey]);
      return newObject;
    }, {});

    return newItem;
  }

  return item;
};

const buildErrorMessage = async (errorObject, reduxState, context) => {
  const {
    user, timestamp, experimentId, url,
  } = context;

  const stack = await StackTrace.fromError(errorObject);
  const stringifiedStack = stack.map((sf) => sf.toString()).join('\n');

  let message = `
    ===== ERROR STACK =====
    Message:
    ${errorObject.message}
    At:
    ${stringifiedStack}

    === DETAILS ===
    User: ${user.attributes.name} <${user.attributes.email}> ${user.username}
    ExperimentID: ${experimentId}
    URL: ${url}
    Timestamp: ${timestamp}

    `;

  if (reduxState) {
    message += `===== REDUX STATE =====
    ${JSON.stringify(reduxState, trimOutput, 2)}`;
  }
  return message;
};

const postError = async (errorLog, context) => {
  const {
    user, timestamp, experimentId, url, environment,
  } = context;

  const message = `
  \u26A0  [${environment.toUpperCase()}] Uncaught UI Error - ExpID ${experimentId} - ${timestamp}
  URL: ${url}

  User: ${user.attributes.name} <${user.attributes.email}> ${user.username}
  Experiment ID: ${experimentId}`;

  const formData = new FormData();

  const formFields = {
    token: getLoggerBotToken(),
    title: `UI Error at ${new Date().toISOString()}`,
    initial_comment: message,
    channels: 'error-logs',
    filename: `UI-error-${experimentId}-${timestamp}.log`,
    content: errorLog,
  };

  Object.entries(formFields).forEach(([fieldName, value]) => {
    formData.append(fieldName, value);
  });

  try {
    const res = await fetch('https://slack.com/api/files.upload', {
      method: 'POST',
      body: formData,
    });

    if (!res.ok) {
      throw new Error(
        `Failed sending error message to Slack: ${res.status}`, res.statusText,
      );
    }
  } catch (err) {
    console.error(err);
  }
};

const postErrorToSlack = async (errorObject, reduxState) => {
  const user = await Auth.currentAuthenticatedUser();

  const timestamp = new Date().toISOString();
  const url = window.location.href;
  const experimentId = extractExperimentId(url);
  const { networkResources: { environment } } = reduxState;

  const context = {
    user,
    timestamp,
    experimentId,
    url,
    environment,
  };

  const errorLog = await buildErrorMessage(errorObject, reduxState, context);
  await postError(errorLog, context);
};

export default postErrorToSlack;
