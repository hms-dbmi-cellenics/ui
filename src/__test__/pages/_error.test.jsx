import React from 'react';
import Error from 'pages/_error';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import postErrorToSlack from 'utils/postErrorToSlack';

import createTestComponentFactory from '__test__/test-utils/testComponentFactory';
import initialState from 'redux/reducers/cellSets/initialState';

jest.mock('utils/postErrorToSlack');

const mockStore = configureMockStore([thunk]);

const mockErrorProp = {
  errorObject: { message: 'Some error message' },
};

const storeState = {
  // Added cellSets initial state to test for dumping of state
  cellSets: initialState,
  networkResources: { environment: 'test' },
};

const defaultProps = {};
const renderErrorFactory = createTestComponentFactory(Error, defaultProps);

const renderErrorPage = (newProps, store = storeState) => {
  render(
    <Provider store={mockStore(store)}>
      {renderErrorFactory(newProps)}
    </Provider>,
  );
};

delete window.location;
window.location = { reload: jest.fn() };

describe('ErrorPage', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Renders properly without props', () => {
    renderErrorPage();

    expect(screen.getByText(/Sorry, something went wrong on our end./i)).toBeInTheDocument();

    // There should be a feedback button and a reload button
    expect(screen.getByText(/Feedback or issues?/i).closest('button')).toBeInTheDocument();
    expect(screen.getByText(/Reload page/i)).toBeInTheDocument();
  });

  it('Shows the HTTP status code and error text', () => {
    renderErrorPage({
      statusCode: 404,
      errorText: 'Not Found',
    });

    expect(screen.getByText(/404/)).toBeInTheDocument();

    expect(screen.getByText(/The error is reported as/i)).toBeInTheDocument();
    expect(screen.getByText(/Not Found/i)).toBeInTheDocument();
  });

  it('Clicking the reload button should reload the page', () => {
    renderErrorPage();

    userEvent.click(screen.getByText(/Reload page/i));

    expect(window.location.reload).toHaveBeenCalledTimes(1);
  });

  it.only('Should post error to Slack if environment is production', () => {
    const productionState = {
      ...storeState,
      networkResources: { environment: 'staging' },
    };

    renderErrorPage(mockErrorProp, productionState);

    expect(postErrorToSlack).toHaveBeenCalledTimes(1);
    expect(postErrorToSlack.mock.calls[0]).toMatchSnapshot();
  });

  it('Should not post error to Slack if environment is not production', () => {
    const stagingState = {
      ...storeState,
      networkResources: { environment: 'production' },
    };

    renderErrorPage(mockErrorProp, stagingState);

    expect(postErrorToSlack).not.toHaveBeenCalled();
  });
});
