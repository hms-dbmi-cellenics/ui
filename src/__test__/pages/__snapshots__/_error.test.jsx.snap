// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorPage Should post error to Slack if environment is production 1`] = `
[
  {
    "message": "Some error message",
  },
  {
    "backendStatus": {},
    "cellInfo": {
      "focus": {
        "key": "louvain",
        "store": "cellSets",
      },
      "groupedTrack": "louvain",
      "selectedTracks": [
        "louvain",
      ],
    },
    "cellMeta": {
      "doubletScores": {
        "data": [],
        "error": false,
        "loading": true,
      },
      "mitochondrialContent": {
        "data": [],
        "error": false,
        "loading": true,
      },
      "numOfGenes": {
        "data": [],
        "error": false,
        "loading": true,
      },
      "numOfUmis": {
        "data": [],
        "error": false,
        "loading": true,
      },
    },
    "cellSets": {
      "error": false,
      "hidden": Set {},
      "hierarchy": [],
      "initialLoadPending": true,
      "loading": false,
      "properties": {},
      "selected": [],
      "updatingClustering": false,
    },
    "componentConfig": {},
    "differentialExpression": {
      "comparison": {
        "advancedFilters": [],
        "group": {
          "between": {
            "basis": null,
            "cellSet": null,
            "compareWith": null,
          },
          "within": {
            "basis": null,
            "cellSet": null,
            "compareWith": null,
          },
        },
        "ordering": {
          "orderBy": "logFC",
          "orderDirection": "DESC",
        },
        "type": "within",
      },
      "properties": {
        "comparisonGroup": {},
        "comparisonType": null,
        "data": [],
        "error": false,
        "loading": false,
        "total": 0,
      },
    },
    "embeddings": {},
    "experimentSettings": {
      "info": {
        "experimentId": null,
        "experimentName": null,
        "pipelineVersion": null,
        "sampleIds": [],
      },
      "originalProcessing": {},
      "processing": {
        "meta": {
          "changedQCFilters": Set {},
          "completingStepError": false,
          "loading": false,
          "loadingSettingsError": false,
        },
      },
    },
    "experiments": {
      "ids": [],
      "meta": {
        "activeExperimentId": null,
        "error": false,
        "loading": false,
        "saving": false,
      },
    },
    "genes": {
      "expression": {
        "downsampled": {
          "ETag": null,
          "cellOrder": [],
          "error": false,
          "loading": [],
          "matrix": ExpressionMatrix {
            "geneIndexes": {},
            "rawGeneExpressions": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
            "setGeneExpression": [Function],
            "stats": {
              "rawMean": [],
              "rawStdev": [],
              "truncatedMax": [],
              "truncatedMin": [],
            },
            "truncatedGeneExpressions": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
            "zScore": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
          },
        },
        "full": {
          "error": false,
          "loading": [],
          "matrix": ExpressionMatrix {
            "geneIndexes": {},
            "rawGeneExpressions": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
            "setGeneExpression": [Function],
            "stats": {
              "rawMean": [],
              "rawStdev": [],
              "truncatedMax": [],
              "truncatedMin": [],
            },
            "truncatedGeneExpressions": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
            "zScore": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
          },
        },
        "views": {},
      },
      "focused": undefined,
      "markers": {
        "ETag": null,
        "error": false,
        "loading": false,
      },
      "properties": {
        "data": {},
        "loading": [],
        "views": {},
      },
      "selected": [],
    },
    "layout": {
      "panel": "Gene list",
      "windows": {
        "direction": "row",
        "first": {
          "direction": "column",
          "first": {
            "direction": "row",
            "first": "UMAP",
            "second": "Cell sets and Metadata",
            "splitPercentage": 60,
          },
          "second": "Heatmap",
          "splitPercentage": 50,
        },
        "second": "Genes",
        "splitPercentage": 70,
      },
    },
    "networkResources": {
      "domainName": "cellenics.hms.harvard.edu",
      "environment": "production",
    },
    "samples": {
      "meta": {
        "error": false,
        "loading": false,
        "saving": false,
        "validating": [],
      },
    },
    "user": {
      "current": null,
    },
  },
]
`;

exports[`ErrorPage Should post error to Slack if environment is staging 1`] = `
[
  {
    "message": "Some error message",
  },
  {
    "backendStatus": {},
    "cellInfo": {
      "focus": {
        "key": "louvain",
        "store": "cellSets",
      },
      "groupedTrack": "louvain",
      "selectedTracks": [
        "louvain",
      ],
    },
    "cellMeta": {
      "doubletScores": {
        "data": [],
        "error": false,
        "loading": true,
      },
      "mitochondrialContent": {
        "data": [],
        "error": false,
        "loading": true,
      },
      "numOfGenes": {
        "data": [],
        "error": false,
        "loading": true,
      },
      "numOfUmis": {
        "data": [],
        "error": false,
        "loading": true,
      },
    },
    "cellSets": {
      "error": false,
      "hidden": Set {},
      "hierarchy": [],
      "initialLoadPending": true,
      "loading": false,
      "properties": {},
      "selected": [],
      "updatingClustering": false,
    },
    "componentConfig": {},
    "differentialExpression": {
      "comparison": {
        "advancedFilters": [],
        "group": {
          "between": {
            "basis": null,
            "cellSet": null,
            "compareWith": null,
          },
          "within": {
            "basis": null,
            "cellSet": null,
            "compareWith": null,
          },
        },
        "ordering": {
          "orderBy": "logFC",
          "orderDirection": "DESC",
        },
        "type": "within",
      },
      "properties": {
        "comparisonGroup": {},
        "comparisonType": null,
        "data": [],
        "error": false,
        "loading": false,
        "total": 0,
      },
    },
    "embeddings": {},
    "experimentSettings": {
      "info": {
        "experimentId": null,
        "experimentName": null,
        "pipelineVersion": null,
        "sampleIds": [],
      },
      "originalProcessing": {},
      "processing": {
        "meta": {
          "changedQCFilters": Set {},
          "completingStepError": false,
          "loading": false,
          "loadingSettingsError": false,
        },
      },
    },
    "experiments": {
      "ids": [],
      "meta": {
        "activeExperimentId": null,
        "error": false,
        "loading": false,
        "saving": false,
      },
    },
    "genes": {
      "expression": {
        "downsampled": {
          "ETag": null,
          "cellOrder": [],
          "error": false,
          "loading": [],
          "matrix": ExpressionMatrix {
            "geneIndexes": {},
            "rawGeneExpressions": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
            "setGeneExpression": [Function],
            "stats": {
              "rawMean": [],
              "rawStdev": [],
              "truncatedMax": [],
              "truncatedMin": [],
            },
            "truncatedGeneExpressions": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
            "zScore": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
          },
        },
        "full": {
          "error": false,
          "loading": [],
          "matrix": ExpressionMatrix {
            "geneIndexes": {},
            "rawGeneExpressions": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
            "setGeneExpression": [Function],
            "stats": {
              "rawMean": [],
              "rawStdev": [],
              "truncatedMax": [],
              "truncatedMin": [],
            },
            "truncatedGeneExpressions": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
            "zScore": {
              "datatype": undefined,
              "index": [],
              "mathjs": "SparseMatrix",
              "ptr": [
                0,
              ],
              "size": [
                0,
                0,
              ],
              "values": [],
            },
          },
        },
        "views": {},
      },
      "focused": undefined,
      "markers": {
        "ETag": null,
        "error": false,
        "loading": false,
      },
      "properties": {
        "data": {},
        "loading": [],
        "views": {},
      },
      "selected": [],
    },
    "layout": {
      "panel": "Gene list",
      "windows": {
        "direction": "row",
        "first": {
          "direction": "column",
          "first": {
            "direction": "row",
            "first": "UMAP",
            "second": "Cell sets and Metadata",
            "splitPercentage": 60,
          },
          "second": "Heatmap",
          "splitPercentage": 50,
        },
        "second": "Genes",
        "splitPercentage": 70,
      },
    },
    "networkResources": {
      "domainName": "cellenics.hms.harvard.edu",
      "environment": "staging",
    },
    "samples": {
      "meta": {
        "error": false,
        "loading": false,
        "saving": false,
        "validating": [],
      },
    },
    "user": {
      "current": null,
    },
  },
]
`;
